### PR TITLE
Refactor URL component extraction

### PR DESCRIPTION
**Pull request by karwa**
This is currently split across 4 functions (well, 3 functions and an initialiser). It's much easier to understand how AHC understands these URL formats if it happens in a single function (especially since schemes like `http+unix` are not standardised).

Additionally, this rejects empty HTTP hostnames, as they are invalid per the spec, and empty socket paths, since they are meaningless.

### CODE REVIEW

**Comment by dnadoba** [comment]
@swift-server-bot test this please

**Review by Lukasa — CHANGES_REQUESTED** [review]
Yup, this makes sense, thanks. Do you mind very briefly adding tests that validate that empty socket paths are now rejected?

**Comment by Lukasa** [comment]
@swift-server-bot add to allowlist

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 1]
```diff
@@ -267,12 +247,7 @@ extension HTTPClient {
             guard let scheme = url.scheme?.lowercased() else {
                 throw HTTPClientError.emptyScheme
             }
-
-            self.kind = try Kind(forScheme: scheme)
-            self.host = try self.kind.hostFromURL(url)
-            self.socketPath = try self.kind.socketPathFromURL(url)
-            self.uri = self.kind.uriFromURL(url)
-
+            (self.kind, self.host, self.socketPath, self.uri) = try Self.deconstructURL(url)
```
Comment (dnadoba) [inline Sources/AsyncHTTPClient/HTTPHandler.swift:250]: Great work! Thank you!

Would you mind returning the `scheme` and `port` in `deconstructURL(_:)` as well?
`scheme` is already a stored property in `HTTPClient.Request` but `port` not. 
https://github.com/swift-server/async-http-client/blob/5dfbba8bb569c0eb98b927d139e81644e344703e/Sources/AsyncHTTPClient/HTTPHandler.swift#L266-L268
However, I don't mind adding `port` as another stored property. This would allow us to better reuse this functionality in the upcoming new `HTTPClientRequest` for async/await support as described in this proposal: 
https://github.com/swift-server/async-http-client/blob/96872470475f18a5667fbf681e2e0f67ad8a0110/docs/async-await.md#new-httpclientrequest-type

Comment (dnadoba) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:250]: Friendly Ping 🙂
I would like to use your changes soon. Do you have any estimate when you have time to work on this again?

Comment (dnadoba) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:250]: @karwa friendly ping 🙂
I would like to use your changes soon. Do you have any estimate when you have time to work on this again?

Comment (karwa) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:250]: Sure, I'll address the comments now and update the patch

**Comment by karwa** [comment]
OK, I've added the test as @Lukasa requested, and `deconstructURL` now extracts the scheme and port number as @dnadoba requested.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 2]
```diff
@@ -108,63 +108,52 @@ extension HTTPClient {
...
+                return (.unixSocket(kind), scheme, "", url.port ?? defaultPort, socketPath, url.uri)
+            case "unix":
+                let socketPath = url.baseURL?.path ?? url.path
+                let uri = url.baseURL != nil ? url.uri : "/"
+                guard !socketPath.isEmpty else {
+                    throw HTTPClientError.missingSocketPath
                 }
+                return (.unixSocket(.baseURL), scheme, "", url.port ?? 80, socketPath, uri)
+            default:
+                throw HTTPClientError.unsupportedScheme(url.scheme!)
```
Comment (dnadoba) [inline Sources/AsyncHTTPClient/HTTPHandler.swift:156]:
```suggestion
                throw HTTPClientError.unsupportedScheme(scheme)
```
We have a non-optional scheme already in scope, no need to force unwrap.

**Review by dnadoba — CHANGES_REQUESTED** [review]
Thanks a lot! One small change and we are good to go from my side. 
CI is currently flaky because I turned parallel testing on. Fix is on its way (https://github.com/swift-server/async-http-client/pull/498) so don't worry if you see any test in `HTTPClientSOCKSTests` failing.

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 2 above)
Comment (karwa) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:156]: The non-optional scheme is lowercased, and I figured it would preserve existing behaviour better to retain the original scheme verbatim. But I don't mind changing it if you think that isn't worth doing.

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 3]
```diff
@@ -215,6 +215,14 @@ class HTTPClientTests: XCTestCase {
         XCTAssertNil(url10)
     }
 
+    func testBadUnixWithBaseURL() {
+        let badUnixBaseURL = URL(string: "/foo", relativeTo: URL(string: "unix:")!)!
+        XCTAssertEqual(badUnixBaseURL.baseURL?.path, "")
+        XCTAssertThrowsError(try Request(url: badUnixBaseURL), "should throw") { error in
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:221]: Smallest nit ever seen: The message doesn't add any information.
```suggestion
        XCTAssertThrowsError(try Request(url: badUnixBaseURL)) { error in
```

**Review by fabianfett — APPROVED** [review]
This is a great change! Thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 2 above)
Comment (dnadoba) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:156]: Makes sense, let's keep it this way.

**Review by dnadoba — APPROVED** [review]

**Review by Lukasa — APPROVED** [review]

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift (hunk 3 above)
Comment (karwa) [reply Tests/AsyncHTTPClientTests/HTTPClientTests.swift:221]: Sure, this is how the ones in `testBadRequestURI` are, but you're right that it doesn't add anything.
